File: src/data/v2/notes.ts
/**
 * 主文档（通常是 `note.md`）的按需读写与冲突保护
 *
 * 规则（设计 §5.3、§14.2）：
 * - 正文是**磁盘上的普通文件**，只有进入节点时才按需读它，绝不随图快照载入；
 * - 保存必须携带手上那份的文档修订号；修订号过期、磁盘哈希不符、文件被删，
 *   都返回结构化冲突（`revision_mismatch` / `hash_mismatch` / `disk_missing`），
 *   让用户在「重新载入 / 覆盖保存 / 另存冲突副本」之间选择，**默认绝不覆盖**；
 * - 「覆盖保存」会先把磁盘上的旧版本另存为冲突副本（放在节点自己的
 *   `.meta/knowledgenet/conflicts/` 里，跟着节点文件夹一起走），用户不会丢东西。
 *
 * 文档修订号**不写进 `node.json`**，而是记在 `.meta/knowledgenet/notes-index.json`
 * （与 Rust 端 `docs/v2-deviations.md` D5 同构）：
 * `node.json` 的 `revision` 描述节点元数据的修订，文档修订描述正文文件的修订；
 * 用一个数字表示两件事，会出现「改了标题导致笔记保存报冲突」这种莫名其妙的交互。
 * 记账文件坏了直接重建——它是可推导的缓存，不是权威数据。
 */
import { RepositoryError } from "../errors.ts";
import type { NodeNote, NoteConflict, NoteDiskState } from "../types.ts";
import { fileEntry, writeTextAtomic, type Vfs } from "./fs.ts";
import { byteLengthOf, sha256Hex } from "./hash.ts";
import { readNodeMeta, readNodeMetaIfExists } from "./nodeMeta.ts";
import {
  emptyNotesIndex,
  parseNotesIndexFile,
  serializeJson,
  type V2NoteEntry,
  type V2NotesIndexFile,
} from "./schema.ts";
import { META_NS_DIR, conflictsDir, extNameOf, joinRel, stemOf } from "./paths.ts";

/** 没有 `primaryDocument` 时的默认正文文件名 */
export const DEFAULT_DOCUMENT = "note.md";

/** 记账文件相对节点目录的路径 */
export const NOTES_INDEX_FILE = `${META_NS_DIR}/notes-index.json`;

export interface NoteFingerprint {
  /** 文档相对节点目录的路径 */
  relativePath: string;
  revision: number;
  sha256: string;
  byteLength: number;
  modifiedAt: number;
}

export interface DocumentSnapshot {
  nodeId: string;
  /** 文档相对节点目录的路径 */
  relativePath: string;
  content: string;
  sha256: string;
  byteLength: number;
  modifiedAt: number;
  documentRevision: number;
}

export type DocumentWriteOutcome =
  | { status: "saved"; note: DocumentSnapshot; conflictCopy: string | null }
  | { status: "conflict"; conflict: NoteConflict; conflictCopy: string | null };

/** 节点当前的主文档路径：元数据里写了就用它，否则用 `note.md` */
export function documentPathOf(primaryDocument: string | null | undefined): string {
  return primaryDocument && primaryDocument.trim() !== "" ? primaryDocument : DEFAULT_DOCUMENT;
}

export function makeNoteFingerprint(input: {
  relativePath: string;
  revision: number;
  content: string;
  modifiedAt?: number;
}): NoteFingerprint {
  return {
    relativePath: input.relativePath,
    revision: input.revision,
    sha256: sha256Hex(input.content),
    byteLength: byteLengthOf(input.content),
    modifiedAt: input.modifiedAt ?? Date.now(),
  };
}

/* ------------------------------ 记账文件读写 ------------------------------ */

function notesIndexPath(nodeRel: string): string {
  return joinRel(nodeRel, NOTES_INDEX_FILE);
}

/** 读记账文件；缺失或坏掉都返回空索引（重建它不需要用户做任何事） */
export async function readNotesIndex(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
): Promise<V2NotesIndexFile> {
  const rel = notesIndexPath(nodeRel);
  try {
    const file = parseNotesIndexFile(await vfs.read(rel), rel);
    return file.nodeId === nodeId ? file : { ...file, nodeId };
  } catch {
    return emptyNotesIndex(nodeId);
  }
}

async function writeNotesIndex(
  vfs: Vfs,
  nodeRel: string,
  file: V2NotesIndexFile,
): Promise<void> {
  await writeTextAtomic(vfs, notesIndexPath(nodeRel), serializeJson(file));
}

/** 取出某个文档的记账项（没有就是 null：说明这份文档还没被本应用登记过） */
export async function fingerprintOf(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
  documentRel: string,
): Promise<NoteFingerprint | null> {
  const file = await readNotesIndex(vfs, nodeRel, nodeId);
  const entry = file.entries.find((item) => item.relativePath === documentRel);
  return entry ? { ...entry } : null;
}

/** 写入/更新一条记账项（保留其它文档的记账） */
async function putFingerprint(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
  fingerprint: NoteFingerprint,
): Promise<void> {
  const file = await readNotesIndex(vfs, nodeRel, nodeId);
  const entry: V2NoteEntry = { ...fingerprint };
  const entries = [...file.entries.filter((item) => item.relativePath !== entry.relativePath), entry];
  await writeNotesIndex(vfs, nodeRel, { ...file, nodeId, entries });
}

/* -------------------------------- 读与检查 -------------------------------- */

async function diskSnapshot(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
  documentRel: string,
  fingerprint: NoteFingerprint | null,
): Promise<DocumentSnapshot> {
  const rel = joinRel(nodeRel, documentRel);
  const entry = await fileEntry(vfs, rel);
  if (!entry) {
    return {
      nodeId,
      relativePath: documentRel,
      content: "",
      sha256: "",
      byteLength: 0,
      modifiedAt: 0,
      documentRevision: fingerprint?.revision ?? 0,
    };
  }
  const content = await vfs.read(rel);
  return {
    nodeId,
    relativePath: documentRel,
    content,
    sha256: sha256Hex(content),
    byteLength: byteLengthOf(content),
    modifiedAt: entry.modifiedMs,
    documentRevision: fingerprint?.revision ?? 0,
  };
}

/** 读主文档（进入节点时调用）；节点不存在时抛 `node_missing` */
export async function readDocument(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
  documentRel?: string,
): Promise<DocumentSnapshot> {
  const snapshot = await readNodeMeta(vfs, nodeRel);
  const doc = documentRel ?? documentPathOf(snapshot.meta.primaryDocument);
  const fingerprint = await fingerprintOf(vfs, nodeRel, nodeId, doc);
  return diskSnapshot(vfs, nodeRel, nodeId, doc, fingerprint);
}

/** 检查磁盘上的主文档与记账是否一致（窗口重新获得焦点、进入节点时调用） */
export async function checkDocument(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
  documentRel?: string,
): Promise<NoteDiskState> {
  const snapshot = await readNodeMetaIfExists(vfs, nodeRel);
  if (!snapshot) {
    throw new RepositoryError("node_missing", `节点不存在：${nodeRel}`, { relativePath: nodeRel });
  }
  const doc = documentRel ?? documentPathOf(snapshot.meta.primaryDocument);
  const fingerprint = await fingerprintOf(vfs, nodeRel, nodeId, doc);
  const disk = await diskSnapshot(vfs, nodeRel, nodeId, doc, fingerprint);
  const exists = disk.modifiedAt > 0 || disk.sha256 !== "";
  return {
    nodeId,
    exists,
    sha256: disk.sha256,
    byteLength: disk.byteLength,
    modifiedAt: disk.modifiedAt,
    documentRevision: disk.documentRevision,
    changedOnDisk: !fingerprint || !exists || disk.sha256 !== fingerprint.sha256,
  };
}

/* --------------------------------- 写入 --------------------------------- */

/**
 * 写主文档。
 *
 * 判定顺序刻意固定：先看修订号（调用方手上的那份是不是最新），再看哈希
 * （磁盘内容有没有被外部编辑器改过）。这样界面上给出的冲突原因总是最贴近
 * 用户实际做错的那一步，而不是笼统地说「保存失败」。
 */
export async function writeDocument(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
  content: string,
  expectedRevision: number,
  force = false,
  documentRel?: string,
): Promise<DocumentWriteOutcome> {
  const snapshot = await readNodeMeta(vfs, nodeRel);
  const doc = documentRel ?? documentPathOf(snapshot.meta.primaryDocument);
  const fingerprint = await fingerprintOf(vfs, nodeRel, nodeId, doc);
  const disk = await diskSnapshot(vfs, nodeRel, nodeId, doc, fingerprint);
  const exists = disk.modifiedAt > 0 || disk.sha256 !== "";
  const registered = fingerprint !== null;

  let reason: NoteConflict["reason"] | null = null;
  let detail = "";
  if (expectedRevision > 0) {
    const actual = fingerprint?.revision ?? 0;
    if (!registered || expectedRevision !== actual) {
      reason = "revision_mismatch";
      detail = `手上这份不是最新版本：期望修订号 ${expectedRevision}，磁盘上是 ${actual}`;
    }
  }
  if (!reason && registered) {
    if (!exists) {
      reason = "disk_missing";
      detail = `磁盘上的 ${doc} 已经不在了（可能被外部删除或移动）`;
    } else if (disk.sha256 !== fingerprint?.sha256) {
      reason = "hash_mismatch";
      detail = `磁盘上的 ${doc} 已被外部修改（内容哈希不一致）`;
    }
  }

  if (reason && !force) {
    return {
      status: "conflict",
      conflict: {
        disk,
        expectedRevision,
        reason,
        detail: `${detail}。可以选择重新载入磁盘版本、另存冲突副本，或明确覆盖。`,
      },
      conflictCopy: null,
    };
  }

  // 覆盖保存：先把磁盘上的旧版本留一份副本，用户不会丢东西
  let conflictCopy: string | null = null;
  if (reason && force && exists) {
    const previousRevision = fingerprint?.revision ?? 0;
    const name = `${stemOf(doc)}-${previousRevision}${extNameOf(doc)}`;
    const copyRel = joinRel(conflictsDir(nodeRel), name);
    await writeTextAtomic(vfs, copyRel, disk.content);
    conflictCopy = copyRel;
  }

  const now = Date.now();
  const nextRevision = (fingerprint?.relativePath ?? "") === doc ? (fingerprint?.revision ?? 0) + 1 : 1;
  const nextFingerprint = makeNoteFingerprint({
    relativePath: doc,
    revision: Math.max(1, nextRevision),
    content,
    modifiedAt: now,
  });
  await writeTextAtomic(vfs, joinRel(nodeRel, doc), content);
  await putFingerprint(vfs, nodeRel, nodeId, nextFingerprint);

  return {
    status: "saved",
    note: {
      nodeId,
      relativePath: doc,
      content,
      sha256: nextFingerprint.sha256,
      byteLength: nextFingerprint.byteLength,
      modifiedAt: now,
      documentRevision: nextFingerprint.revision,
    },
    conflictCopy,
  };
}

/**
 * 新建节点时写一份初始主文档（桌面端与演示端都这么做）。
 *
 * 为什么新建节点一定要有 `note.md`：用户点开节点就能直接写，
 * 而不是先面对一句「还没有正文，是否创建」。
 */
export async function createPrimaryDocument(
  vfs: Vfs,
  nodeRel: string,
  nodeId: string,
  content = "",
  documentRel: string = DEFAULT_DOCUMENT,
): Promise<{ snapshot: DocumentSnapshot; documentRevision: number }> {
  await writeTextAtomic(vfs, joinRel(nodeRel, documentRel), content);
  const fingerprint = makeNoteFingerprint({ relativePath: documentRel, revision: 1, content });
  await putFingerprint(vfs, nodeRel, nodeId, fingerprint);
  return {
    snapshot: {
      nodeId,
      relativePath: documentRel,
      content,
      sha256: fingerprint.sha256,
      byteLength: fingerprint.byteLength,
      modifiedAt: fingerprint.modifiedAt,
      documentRevision: 1,
    },
    documentRevision: 1,
  };
}

/** 列出节点里的冲突副本（界面「查看被覆盖的旧版本」用） */
export async function listConflictCopies(
  vfs: Vfs,
  nodeRel: string,
): Promise<Array<{ relativePath: string; name: string; content: string }>> {
  const dir = conflictsDir(nodeRel);
  let entries;
  try {
    entries = await vfs.list(dir);
  } catch {
    return [];
  }
  const out: Array<{ relativePath: string; name: string; content: string }> = [];
  for (const entry of entries) {
    if (entry.kind !== "file") continue;
    const rel = joinRel(dir, entry.name);
    out.push({ relativePath: rel, name: entry.name, content: await vfs.read(rel) });
  }
  return out;
}

/** 把文档快照转成仓储层返回的形状（领域模型与快照字段一致，这里只是集中一处） */
export function toNodeNote(snapshot: DocumentSnapshot): NodeNote {
  return {
    nodeId: snapshot.nodeId,
    relativePath: snapshot.relativePath,
    content: snapshot.content,
    documentRevision: snapshot.documentRevision,
    sha256: snapshot.sha256,
    byteLength: snapshot.byteLength,
    modifiedAt: snapshot.modifiedAt,
  };
}

File: src-tauri/src/v2/notes.rs
//! 主文档（`primaryDocument`）读写与冲突保护。
//!
//! 契约 `docs/v2-contract.md` §3.8。
//!
//! **文档修订号不进 `node.json`**。`node.json` 是节点的身份与元数据，正文属于用户文件，
//! 两者被外部编辑器改动的频率完全不同；把正文修订号塞进 `node.json` 会让「用记事本
//! 改一句笔记」变成「改节点元数据」，也会让每个节点多一次元数据写入。
//!
//! 这里改用节点自己的记账文件
//! `<节点>/.meta/knowledgenet/notes-index.json`（格式见 [`NOTES_INDEX_FORMAT`]）：
//! 每个文档一条「上次已知哈希 + 字节数 + 修订号 + 时间」。修订号是乐观并发用的计数器，
//! 哈希是判断「磁盘上这份内容是不是我上次见过的那份」的依据。索引丢失或损坏时
//! 修订号退回 0（未知），下次保存会要求用户重新载入，而不是拿一个可能过期的版本去覆盖。

use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::models::{code, CmdError, CmdResult};
use crate::paths;

use super::atomic::{self, Fingerprint};
use super::ctx::LibCtx;
use super::nodes::{self, NodeMetaPatch};
use super::schema::{validate_primary_document, NodeMeta, Validate};
use super::vpaths;

/* --------------------------------- 常量 --------------------------------- */

/// 记账文件在 `.meta/knowledgenet/` 下的文件名。
pub const NOTES_INDEX_FILE: &str = "notes-index.json";
/// `notes-index.json` 的格式标记（自描述，便于人工识别与以后升级）。
pub const NOTES_INDEX_FORMAT: &str = "knowledgenet-notes-index";
pub const NOTES_INDEX_FORMAT_VERSION: i64 = 1;
/// `primaryDocument` 为空时，第一次保存落在哪个文件。
pub const DEFAULT_DOCUMENT: &str = "note.md";

/* --------------------------------- 模型 --------------------------------- */

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NotesIndexEntry {
    pub relative_path: String,
    #[serde(default)]
    pub sha256: String,
    #[serde(default)]
    pub byte_length: i64,
    #[serde(default)]
    pub revision: i64,
    #[serde(default)]
    pub modified_at: String,
    #[serde(flatten)]
    pub extra: serde_json::Map<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NotesIndex {
    pub format: String,
    pub format_version: i64,
    pub node_id: String,
    #[serde(default)]
    pub entries: Vec<NotesIndexEntry>,
    #[serde(flatten)]
    pub extra: serde_json::Map<String, serde_json::Value>,
}

impl NotesIndex {
    pub fn empty(node_id: &str) -> Self {
        Self {
            format: NOTES_INDEX_FORMAT.to_string(),
            format_version: NOTES_INDEX_FORMAT_VERSION,
            node_id: node_id.to_string(),
            entries: Vec::new(),
            extra: serde_json::Map::new(),
        }
    }

    pub fn entry(&self, relative_path: &str) -> Option<&NotesIndexEntry> {
        self.entries.iter().find(|e| e.relative_path == relative_path)
    }

    pub fn upsert(&mut self, entry: NotesIndexEntry) {
        match self
            .entries
            .iter_mut()
            .find(|e| e.relative_path == entry.relative_path)
        {
            Some(existing) => *existing = entry,
            None => self.entries.push(entry),
        }
        self.entries.sort_by(|a, b| a.relative_path.cmp(&b.relative_path));
    }
}

/// 记账文件没有「必须拒绝」的字段：格式头由 `atomic::read_typed` 先校验，
/// 剩下的内容是修订号账本，读不出来时按「未知」处理而不是让笔记读写失败。
impl Validate for NotesIndex {
    fn validate_typed(&self) -> CmdResult<()> {
        Ok(())
    }
}

/// 一份文档在某个瞬间的完整快照。
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DocumentSnapshot {
    /// 节点目录内的相对路径；`primaryDocument` 为空且文件不存在时是空串
    pub relative_path: String,
    pub content: String,
    #[serde(default)]
    pub sha256: String,
    #[serde(default)]
    pub byte_length: i64,
    #[serde(default)]
    pub modified_ms: i64,
    /// 乐观并发用的文档修订号；0 表示「未知 / 还没有这份文档」
    #[serde(default)]
    pub revision: i64,
}

impl DocumentSnapshot {
    fn empty(relative_path: &str) -> Self {
        Self {
            relative_path: relative_path.to_string(),
            content: String::new(),
            sha256: String::new(),
            byte_length: 0,
            modified_ms: 0,
            revision: 0,
        }
    }

    pub fn is_empty_document(&self) -> bool {
        self.relative_path.is_empty()
    }
}

/// 写文档的结果。冲突是**正常返回值**而不是异常：界面必须能拿到磁盘版本，
/// 让用户选「重新载入 / 覆盖保存（自动留副本）」。
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", rename_all_fields = "camelCase", tag = "status")]
pub enum WriteDocumentOutcome {
    Saved {
        document: DocumentSnapshot,
        /// 「覆盖保存」时磁盘旧版本被另存的位置（相对知识库根）
        #[serde(default)]
        conflict_copy: Option<String>,
    },
    Conflict {
        disk: DocumentSnapshot,
        expected_revision: i64,
        /// `revision_mismatch` | `external_edit` | `disk_missing`
        reason: String,
        detail: String,
    },
}

/* ------------------------------ 记账文件读写 ------------------------------ */

struct IndexState {
    index: NotesIndex,
    /// 记录可信（修订号可用）
    trusted: bool,
    /// 可以覆盖写（损坏的文件可以重建；更高版本的别人家的文件不动）
    replaceable: bool,
}

fn index_path(node_dir: &Path) -> PathBuf {
    vpaths::ns_dir(node_dir).join(NOTES_INDEX_FILE)
}

fn load_index(ctx: &LibCtx, node_dir: &Path, node_id: &str) -> IndexState {
    let path = index_path(node_dir);
    let empty = NotesIndex::empty(node_id);
    if !path.is_file() {
        return IndexState {
            index: empty,
            trusted: true,
            replaceable: true,
        };
    }
    let relative = nodes::rel_of(ctx, &path);
    match atomic::read_typed::<NotesIndex>(
        &path,
        NOTES_INDEX_FORMAT,
        NOTES_INDEX_FORMAT_VERSION,
        "notes-index.json",
        Some(&relative),
    ) {
        Ok((mut index, _fp)) => {
            index.node_id = node_id.to_string();
            IndexState {
                index,
                trusted: true,
                replaceable: true,
            }
        }
        Err(err) if err.code == code::METADATA_UNSUPPORTED => {
            // 更高版本的记账文件：不认识就不动它，修订号按「未知」处理
            eprintln!("{relative} 的版本不受支持，本次不改写它：{}", err.message);
            IndexState {
                index: empty,
                trusted: false,
                replaceable: false,
            }
        }
        Err(err) => {
            eprintln!("{relative} 无法解析，将按未知修订号处理：{}", err.message);
            IndexState {
                index: empty,
                trusted: false,
                replaceable: true,
            }
        }
    }
}

fn store_index(path: &Path, index: &NotesIndex) -> CmdResult<()> {
    let mut value = index.clone();
    value.entries.sort_by(|a, b| a.relative_path.cmp(&b.relative_path));
    atomic::write_json(path, &value).map(|_| ())
}

/* --------------------------------- 读取 --------------------------------- */

/// 决定这次读写落在哪个文件：显式给路径就用它，否则用 `node.json` 的 `primaryDocument`。
fn resolve_document_rel(meta: &NodeMeta, document_rel: &str) -> CmdResult<Option<String>> {
    let explicit = document_rel.trim();
    if !explicit.is_empty() {
        return Ok(Some(validate_primary_document(explicit)?));
    }
    match &meta.primary_document {
        Some(raw) if !raw.trim().is_empty() => Ok(Some(validate_primary_document(raw)?)),
        _ => Ok(None),
    }
}

/// 修订号推导：索引记住了这份内容 → 用它记的修订号；
/// 磁盘内容和索引记的不一样（外部改过 / 从没见过）→ 索引修订号 + 1；
/// 索引不可信（没有 / 损坏 / 高版本）→ 0（未知，必须重新载入）。
fn derive_revision(state: &IndexState, relative_path: &str, disk_sha256: &str) -> i64 {
    if !state.trusted {
        return 0;
    }
    match state.index.entry(relative_path) {
        Some(entry) if entry.sha256.eq_ignore_ascii_case(disk_sha256) && !disk_sha256.is_empty() => {
            entry.revision.max(0)
        }
        Some(entry) => entry.revision.max(0) + 1,
        None => 1,
    }
}

fn snapshot_at(
    ctx: &LibCtx,
    node_dir: &Path,
    relative_path: &str,
    state: &IndexState,
) -> CmdResult<DocumentSnapshot> {
    let path = document_path(node_dir, relative_path)?;
    let Some(fp): Option<Fingerprint> = atomic::fingerprint_opt(&path)? else {
        return Ok(DocumentSnapshot::empty(relative_path));
    };
    let content = paths::read_text(&path)?;
    let _ = ctx;
    Ok(DocumentSnapshot {
        relative_path: relative_path.to_string(),
        content,
        sha256: fp.sha256.clone(),
        byte_length: fp.bytes as i64,
        modified_ms: fp.modified_ms,
        revision: derive_revision(state, relative_path, &fp.sha256),
    })
}

fn document_path(node_dir: &Path, relative_path: &str) -> CmdResult<PathBuf> {
    let cleaned = validate_primary_document(relative_path)?;
    Ok(node_dir.join(cleaned.replace('/', std::path::MAIN_SEPARATOR_STR)))
}

/// 读主文档。
///
/// - `document_rel` 为空时读 `node.json` 的 `primaryDocument`；
/// - `primaryDocument` 也是空时返回空内容 + 修订号 0（保存时会自动落到 `note.md`）；
/// - 登记了主文档但文件不在磁盘上时同样返回空内容 + 修订号 0（保存能重新创建它）。
pub fn read_document(
    ctx: &LibCtx,
    node_rel: &str,
    document_rel: &str,
) -> CmdResult<DocumentSnapshot> {
    let node_dir = ctx.node_dir(node_rel)?;
    let (meta, _fp) = nodes::read_node_meta(ctx, node_rel)?;
    let Some(target) = resolve_document_rel(&meta, document_rel)? else {
        return Ok(DocumentSnapshot::empty(""));
    };
    let state = load_index(ctx, &node_dir, &meta.id);
    snapshot_at(ctx, &node_dir, &target, &state)
}

/// 命令层 `check_node_note` 用：磁盘上的文档状态 + 「它与我们上次记的不一样」。
///
/// 第二个返回值是 `changed_on_disk`：索引记的哈希与磁盘不符（外部编辑器改过），
/// 或者文件消失了。界面据此在保存前提醒用户。
pub fn check_document(ctx: &LibCtx, node_rel: &str) -> CmdResult<(DocumentSnapshot, bool)> {
    let node_dir = ctx.node_dir(node_rel)?;
    let (meta, _fp) = nodes::read_node_meta(ctx, node_rel)?;
    let state = load_index(ctx, &node_dir, &meta.id);
    let Some(target) = resolve_document_rel(&meta, "")? else {
        return Ok((DocumentSnapshot::empty(""), false));
    };
    let snapshot = snapshot_at(ctx, &node_dir, &target, &state)?;
    let changed = match state.index.entry(&target) {
        Some(entry) => {
            snapshot.sha256.is_empty() || !entry.sha256.eq_ignore_ascii_case(&snapshot.sha256)
        }
        None => !snapshot.sha256.is_empty(),
    };
    Ok((snapshot, changed))
}

/* --------------------------------- 写入 --------------------------------- */

/// 把磁盘上的旧版本另存到 `.knowledgenet/conflicts/<nodeId>/<时间戳>-<文件名>`。
///
/// 放在知识库根下的 `.knowledgenet/` 里而不是节点目录里：冲突副本不是知识资产，
/// 不该混进用户的文件夹；但它也不该被自动删除，用户自己决定何时清理。
fn save_conflict_copy(ctx: &LibCtx, node_id: &str, source: &Path, target_rel: &str) -> CmdResult<String> {
    let dir = vpaths::root_meta_dir(ctx.root())
        .join(paths::CONFLICT_DIR)
        .join(node_id);
    atomic::ensure_dir(&dir)?;
    let stamp = paths::now_ms();
    let name = target_rel
        .rsplit('/')
        .next()
        .filter(|n| !n.is_empty())
        .unwrap_or("document");
    let mut destination = dir.join(format!("{stamp}-{name}"));
    let mut suffix = 1;
    while destination.exists() {
        destination = dir.join(format!("{stamp}-{suffix}-{name}"));
        suffix += 1;
        if suffix > 1000 {
            return Err(CmdError::io(format!(
                "冲突副本目录 {} 里同名文件过多",
                dir.display()
            )));
        }
    }
    paths::copy_stream(source, &destination)?;
    ctx.relative_of(&destination)
}

/// 写主文档（带冲突保护）。
///
/// 三条分支：
///
/// - 修订号与磁盘一致 → 直接保存；
/// - 修订号过期且 `force = false` → 返回 [`WriteDocumentOutcome::Conflict`]，**不写盘**；
/// - 修订号过期且 `force = true` → 先把磁盘旧内容复制到 `.knowledgenet/conflicts/...`
///   再覆盖（覆盖掉的那一版永远找得回来）。
///
/// `primaryDocument` 为空（或与本次写入的文档不一致）时顺带把它登记进 `node.json`
/// （这一步会 bump 节点修订号）；登记发生在写正文**之前**，这样不会出现
/// 「正文写了、节点却还不知道自己有哪些文档」的半截状态。
pub fn write_document(
    ctx: &LibCtx,
    node_rel: &str,
    document_rel: &str,
    content: &str,
    expected_revision: i64,
    force: bool,
) -> CmdResult<WriteDocumentOutcome> {
    ctx.require_writable()?;
    let node_dir = ctx.node_dir(node_rel)?;
    let (meta, meta_fp) = nodes::read_node_meta(ctx, node_rel)?;

    let target = resolve_document_rel(&meta, document_rel)?
        .unwrap_or_else(|| DEFAULT_DOCUMENT.to_string());
    let path = document_path(&node_dir, &target)?;

    let state = load_index(ctx, &node_dir, &meta.id);
    let disk_fp = atomic::fingerprint_opt(&path)?;
    let disk_revision = match &disk_fp {
        Some(fp) => derive_revision(&state, &target, &fp.sha256),
        None => 0,
    };

    // 修订号不符 = 磁盘上的东西不是调用方手上那一份
    let mismatch = expected_revision >= 0 && expected_revision != disk_revision;
    let reason = if !mismatch {
        String::new()
    } else if disk_fp.is_none() && expected_revision > 0 {
        "disk_missing".to_string()
    } else {
        match state.index.entry(&target) {
            Some(entry) => match &disk_fp {
                Some(fp) if entry.sha256.eq_ignore_ascii_case(&fp.sha256) => {
                    "revision_mismatch".to_string()
                }
                _ => "external_edit".to_string(),
            },
            None => "external_edit".to_string(),
        }
    };

    if !reason.is_empty() && !force {
        let disk = snapshot_at(ctx, &node_dir, &target, &state)?;
        let detail = format!(
            "{} 期望修订号 {}，磁盘上是 {}（{} 字节，sha256 {}…）",
            target,
            expected_revision,
            disk_revision,
            disk.byte_length,
            disk.sha256.chars().take(12).collect::<String>()
        );
        return Ok(WriteDocumentOutcome::Conflict {
            disk,
            expected_revision,
            reason,
            detail,
        });
    }

    // 「覆盖保存」：先把被覆盖的那一版留一份副本
    let mut conflict_copy = None;
    if !reason.is_empty() && force && path.is_file() {
        conflict_copy = Some(save_conflict_copy(ctx, &meta.id, &path, &target)?);
    }

    // 先登记主文档（失败时一个正文字节都没动）
    if meta.primary_document.as_deref() != Some(target.as_str()) {
        let patch = NodeMetaPatch {
            primary_document: Some(Some(target.clone())),
            ..Default::default()
        };
        nodes::update_node_meta(ctx, node_rel, &patch, meta.revision, Some(&meta_fp.sha256))?;
    }

    atomic::ensure_dir(path.parent().unwrap_or(&node_dir))?;
    let fp = atomic::write_text(&path, content)?;

    let issued_revision = disk_revision + 1;
    let mut fresh = load_index(ctx, &node_dir, &meta.id);
    if fresh.replaceable {
        fresh.index.node_id = meta.id.clone();
        fresh.index.upsert(NotesIndexEntry {
            relative_path: target.clone(),
            sha256: fp.sha256.clone(),
            byte_length: fp.bytes as i64,
            revision: issued_revision,
            modified_at: paths::iso_from_ms(fp.modified_ms),
            extra: serde_json::Map::new(),
        });
        store_index(&index_path(&node_dir), &fresh.index)?;
    }

    Ok(WriteDocumentOutcome::Saved {
        document: DocumentSnapshot {
            relative_path: target,
            content: content.to_string(),
            sha256: fp.sha256,
            byte_length: fp.bytes as i64,
            modified_ms: fp.modified_ms,
            revision: issued_revision,
        },
        conflict_copy,
    })
}
